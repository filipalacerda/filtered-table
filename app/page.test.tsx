import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";

import datastore from "./data/data";

import Home from "./page";
import userEvent from "@testing-library/user-event";

describe(Home.name, () => {
  let container: HTMLElement;
  const products = datastore.getProducts();

  beforeEach(() => {
    container = render(<Home />).container;
  });

  describe("when component renders", () => {
    it("should match snapshot", () => {
      expect(container).toMatchSnapshot();
    });

    it("should render the category select", () => {
      expect(screen.getByTestId("category-select")).toBeInTheDocument();
    });

    it("should not render the operator selector", () => {
      expect(screen.queryByTestId("operator-selector")).not.toBeInTheDocument();
    });

    it("should not render the value field", () => {
      expect(screen.queryByTestId("text-field")).not.toBeInTheDocument();
      expect(screen.queryByTestId("select-field")).not.toBeInTheDocument();
    });

    it("should render the clear button", () => {
      expect(screen.getByRole("button", { name: "Clear" })).toBeInTheDocument();
    });

    it("should render a table with all the products", () => {
      expect(container.querySelectorAll("tbody tr").length).toEqual(
        products.length
      );
    });
  });

  describe("when the user performs a search action", () => {
    let inputField: HTMLElement;

    beforeEach(async () => {
      const categoryInput = screen.getByTestId("category-select");

      await userEvent.click(categoryInput);

      await userEvent.selectOptions(categoryInput, "Product Name");

      const operatorsField = screen.getByTestId("operator-select");
      await userEvent.click(operatorsField);

      await userEvent.selectOptions(operatorsField, "equals");

      inputField = screen.getByTestId("text-field");
    });

    it("should render only the correct results on the table", async () => {
      await userEvent.type(inputField, "headphones");

      expect(
        screen.getByText(products[0].property_values[0].value)
      ).toBeInTheDocument();

      expect(
        screen.queryByText(products[1].property_values[0].value)
      ).not.toBeInTheDocument();
    });

    it("should render the no results message when no results are found", async () => {
      await userEvent.type(inputField, "no matches");

      expect(
        screen.getByText(
          "No results found! Please try another search combination."
        )
      ).toBeInTheDocument();
    });

    it("should clear the filters when the user clicks on the clear button", async () => {
      await userEvent.type(inputField, "no matches");

      await userEvent.click(screen.getByRole("button", { name: "Clear" }));

      expect(
        screen.getByText(products[0].property_values[0].value)
      ).toBeInTheDocument();
    });
  });
});
