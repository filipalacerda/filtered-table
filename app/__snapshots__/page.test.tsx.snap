// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App when component renders should match snapshot 1`] = `
<div>
  <main
    class="app"
  >
    <section
      class="search-container"
    >
      <form
        class="search-form"
      >
        <div
          class="fields-container"
        >
          <fieldset
            class="fieldset"
          >
            <label
              class="label"
              for="category"
            >
              Category:
            </label>
            <select
              aria-required="true"
              class="input"
              data-testid="category-select"
              id="category"
              name="category"
              required=""
            >
              <option
                value="-1"
              >
                Choose Category
              </option>
              <option
                class="bg-white"
                value="0"
              >
                Product Name
              </option>
              <option
                class="bg-white"
                value="1"
              >
                color
              </option>
              <option
                class="bg-white"
                value="2"
              >
                weight (oz)
              </option>
              <option
                class="bg-white"
                value="3"
              >
                category
              </option>
              <option
                class="bg-white"
                value="4"
              >
                wireless
              </option>
            </select>
          </fieldset>
        </div>
        <div>
          <button
            class="clear-button"
            type="button"
          >
            Clear
          </button>
        </div>
      </form>
    </section>
    <section
      class="table-container"
    >
      <table
        class="table"
      >
        <thead
          class="table-header"
        >
          <tr>
            <th
              class="header-cell"
              scope="col"
            >
              Product Name
            </th>
            <th
              class="header-cell"
              scope="col"
            >
              color
            </th>
            <th
              class="header-cell"
              scope="col"
            >
              weight (oz)
            </th>
            <th
              class="header-cell"
              scope="col"
            >
              category
            </th>
            <th
              class="header-cell"
              scope="col"
            >
              wireless
            </th>
          </tr>
        </thead>
        <tbody>
          <tr
            class="table-row"
          >
            <td
              class="cell"
            >
              Headphones
            </td>
            <td
              class="cell"
            >
              black
            </td>
            <td
              class="cell"
            >
              5
            </td>
            <td
              class="cell"
            >
              electronics
            </td>
            <td
              class="cell"
            >
              false
            </td>
          </tr>
          <tr
            class="table-row"
          >
            <td
              class="cell"
            >
              Cell Phone
            </td>
            <td
              class="cell"
            >
              black
            </td>
            <td
              class="cell"
            >
              3
            </td>
            <td
              class="cell"
            >
              electronics
            </td>
            <td
              class="cell"
            >
              true
            </td>
          </tr>
          <tr
            class="table-row"
          >
            <td
              class="cell"
            >
              Keyboard
            </td>
            <td
              class="cell"
            >
              grey
            </td>
            <td
              class="cell"
            >
              5
            </td>
            <td
              class="cell"
            >
              electronics
            </td>
            <td
              class="cell"
            >
              false
            </td>
          </tr>
          <tr
            class="table-row"
          >
            <td
              class="cell"
            >
              Cup
            </td>
            <td
              class="cell"
            >
              white
            </td>
            <td
              class="cell"
            >
              3
            </td>
            <td
              class="cell"
            >
              kitchenware
            </td>
          </tr>
          <tr
            class="table-row"
          >
            <td
              class="cell"
            >
              Key
            </td>
            <td
              class="cell"
            >
              silver
            </td>
            <td
              class="cell"
            >
              1
            </td>
            <td
              class="cell"
            >
              tools
            </td>
          </tr>
          <tr
            class="table-row"
          >
            <td
              class="cell"
            >
              Hammer
            </td>
            <td
              class="cell"
            >
              brown
            </td>
            <td
              class="cell"
            >
              19
            </td>
            <td
              class="cell"
            >
              tools
            </td>
          </tr>
        </tbody>
      </table>
    </section>
  </main>
</div>
`;
